File: Lg_python/P1451/P1451.py
#!/usr/bin/env python

#左上右下
step = [ [ 0, -1 ],
         [ 1, 0 ],
         [ 0, 1 ],
         [ -1, 0 ] ]

#DFS
def Dfs( a, row, col ):
    global r
    global c
    global book
    #print(book)
    for i in range(4):
        x = row + step[i][0]
        y = col + step[i][1]

        #判断是否越界
        if x > r - 1 or y >  c - 1 or x < 0 or y < 0 :
            continue
        
        #判断是否走过
        if book[x][y] == 1:
            continue

        #记录
        if a[x][y] > '0' :
            book[x][y] = 1
            Dfs(a , x ,y )

    

'=============主函数==========='

#读入行列数
r,c = map( int, input().split() )

#初始化矩阵
a = list()
time = 0
ans = 0

#读入矩阵
for row in range( 1, r + 1 ):
    cb = input()
    b = list()
    for each in cb:
        '''temp = int( each )
        b.append( temp )'''
        b.append(each)
    a.append(b)

#记录是否走过
book = [ [ 0 for i in range( c ) ] for j in range( r ) ]

#开始查找细胞
for row in range( r ):
     for col in range( c ):
        if a[row][col] == '0' or book[row][col] == 1:
            continue
        else:
            Dfs( a, row, col )
            book[row][col] = 1
            #print(row,col)
            ans += 1

#输出答案
print(ans)

File: 洛谷/Lg_cpp/P1451/P1451.cpp
#include <iostream>
#include <cstdio>

using namespace std;

void Dfs( int row, int col );

bool map[201][201] = {false};		//地图 
bool book[201][201] = {false};	//记录是否遍历过
int r, c;		//行列 

int main()
{
	//输入行、列
	cin >> r >> c;
	
	//输入地图 
	for( int i = 1; i <= r; i++ )
		for( int j = 1; j <= c; j++ ) 
		{
			int loca;
			scanf("%1d", &loca);
			
			//大于0记录 
			if( loca > 0 )
				map[i][j] = true;
			else
				map[i][j] = false;
		}
	
	//开始遍历寻找细胞数
int count = 0;	//记录细胞数 
	for( int i = 1; i <= r; i++ )
		for( int j = 1; j <= c; j++ ) 
		{
			//存在细胞且没遍历过 
			if( map[i][j] && !book[i][j] )
			{
				book[i][j] = true;
				Dfs( i, j );
				count++;
			}
		}
	
	cout << count;
    return 0;
}

//右上左下 
int step[5] = { 0, 1, 0, -1, 0}; 
void Dfs( int row, int col )
{
	for( int i = 0; i <= 3; i++ )
	{
		int x = row + step[i];
		int y = col + step[i+1];
		
		//判断是否越界
		if( x < 1 || y < 1 || x > r || y > c )
			continue;
		
		//判断是否遍历过或是否可以是细胞的一部分 
		if( map[x][y] && !book[x][y] )
		{
			book[x][y] = true;
			Dfs( x, y );
		}
	} 
}


